### ISSUE TITLE
All message in journal lost if reroute ES api use on unassigned shard and graylog rotation

### ISSUE THREAD

**Issue by rlanore** [Graylog2/graylog2-server#1581]
Today, graylog lost all message in journal.

use debian package:

```
ii  elasticsearch                            1.7.3
ii  graylog-1.0-repository-debian7           1.2.0-1   
ii  graylog-server                           1.2.2-1
ii  graylog-web                              1.2.2-1
```

graylog index rotation is 1 day

ES cluster change to RED state because of disk watermark

I have extended this disque on node. After use reroute ES api to assigne shard. And journal is cleaned.

But it seem to graylog want to write message to old index ( that is forbidden )

Time based event:
2015-11-22 08:35:05,698  -- Low disk watermark ES
2015-11-22 08:35:35,701  -- High disk watermark ES
2015-11-23T00:00:08.865 -- Graylog index rotation pointing to graylog2_366
2015-11-23 09:21:58,846  -- ES REST API reroute unassigned shard
2015-11-23T09:24:45.833 -- Graylog can't process index

Log Graylog

```
2015-11-23T00:00:08.865Z INFO  [IndexRotationThread] Deflector index <graylog2_365> should be rotated, Pointing deflector to new index now!
2015-11-23T00:00:08.866Z INFO  [Deflector] Cycling deflector to next index now.
2015-11-23T00:00:08.883Z INFO  [Deflector] Cycling from <graylog2_365> to <graylog2_366>
2015-11-23T00:00:08.883Z INFO  [Deflector] Creating index target <graylog2_366>...
2015-11-23T00:00:09.154Z INFO  [Deflector] Waiting for index allocation of <graylog2_366>
2015-11-23T00:00:39.156Z INFO  [Deflector] Done!
2015-11-23T00:00:39.156Z INFO  [Deflector] Pointing deflector to new target index....
2015-11-23T00:00:39.198Z INFO  [SystemJobManager] Submitted SystemJob <3b032090-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob]
2015-11-23T00:00:39.198Z INFO  [CreateNewSingleIndexRangeJob] Calculating ranges for index graylog2_365.
2015-11-23T00:00:39.199Z INFO  [SystemJobManager] Submitted SystemJob <3b040af0-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.SetIndexReadOnlyJob]
2015-11-23T00:00:39.199Z INFO  [SystemJobManager] Submitted SystemJob <3b040af1-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob]
2015-11-23T00:00:39.199Z INFO  [Deflector] Done!
2015-11-23T00:00:39.200Z INFO  [CreateNewSingleIndexRangeJob] Calculating ranges for index graylog2_366.
2015-11-23T00:00:39.201Z ERROR [Indices] Error while calculating timestamp stats in index <graylog2_366>
org.elasticsearch.action.search.SearchPhaseExecutionException: Failed to execute phase [query], all shards failed
  at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.onFirstPhaseResult(TransportSearchTypeAction.java:237)
  at org.elasticsearch.action.search.type.TransportSearchTypeAction$BaseAsyncAction.start(TransportSearchTypeAction.java:160)
  at org.elasticsearch.action.search.type.TransportSearchCountAction.doExecute(TransportSearchCountAction.java:55)
  at org.elasticsearch.action.search.type.TransportSearchCountAction.doExecute(TransportSearchCountAction.java:45)
  at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:75)
  at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:108)
  at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:43)
  at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:75)
  at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:55)
  at org.elasticsearch.client.node.NodeClient.execute(NodeClient.java:90)
  at org.elasticsearch.client.FilterClient.execute(FilterClient.java:60)
  at org.graylog2.indexer.elasticsearch.GlobalTimeoutClient.execute(GlobalTimeoutClient.java:53)
  at org.elasticsearch.client.support.AbstractClient.search(AbstractClient.java:333)
  at org.graylog2.indexer.indices.Indices.timestampStatsOfIndex(Indices.java:477)
  at org.graylog2.indexer.ranges.MongoIndexRangeService.calculateRange(MongoIndexRangeService.java:118)
  at org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob.execute(CreateNewSingleIndexRangeJob.java:59)
  at org.graylog2.system.jobs.SystemJobManager$1.run(SystemJobManager.java:88)
  at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:235)
  at java.util.concurrent.FutureTask.run(FutureTask.java:262)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
  at java.lang.Thread.run(Thread.java:745)
2015-11-23T00:00:39.202Z ERROR [CreateNewSingleIndexRangeJob] Exception during index range calculation for index graylog2_366
org.elasticsearch.indices.IndexMissingException: [graylog2_366] missing
at org.graylog2.indexer.indices.Indices.timestampStatsOfIndex(Indices.java:482)
  at org.graylog2.indexer.ranges.MongoIndexRangeService.calculateRange(MongoIndexRangeService.java:118)
  at org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob.execute(CreateNewSingleIndexRangeJob.java:59)
  at org.graylog2.system.jobs.SystemJobManager$1.run(SystemJobManager.java:88)
  at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:235)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
  at java.util.concurrent.FutureTask.run(FutureTask.java:262)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
  at java.lang.Thread.run(Thread.java:745)
2015-11-23T00:00:39.203Z INFO  [SystemJobManager] SystemJob <3b040af1-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob] finished in 3ms.
2015-11-23T00:00:39.214Z WARN  [IndexRotationThread] Deflector is pointing to [graylog2_366], not the newest one: [graylog2_365]. Re-pointing.
2015-11-23T00:00:42.569Z INFO  [MongoIndexRangeService] Calculated range of [graylog2_365] in [3370ms].
2015-11-23T00:00:42.576Z INFO  [CreateNewSingleIndexRangeJob] Created ranges for index graylog2_365.
2015-11-23T00:00:42.576Z INFO  [SystemJobManager] SystemJob <3b032090-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.ranges.CreateNewSingleIndexRangeJob] finished in 3377ms.
2015-11-23T00:01:09.199Z INFO  [SetIndexReadOnlyJob] Flushing old index <graylog2_365>.
2015-11-23T00:01:09.326Z INFO  [SetIndexReadOnlyJob] Setting old index <graylog2_365> to read-only.
2015-11-23T00:01:09.394Z INFO  [SystemJobManager] Submitted SystemJob <4d036f20-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.indices.jobs.OptimizeIndexJob]
2015-11-23T00:01:09.394Z INFO  [SystemJobManager] SystemJob <3b040af0-9175-11e5-be71-005056a37afa> [org.graylog2.indexer.SetIndexReadOnlyJob] finished in 195ms.
2015-11-23T00:01:09.395Z INFO  [OptimizeIndexJob] Optimizing index <graylog2_365>.
2015-11-23T00:01:32.132Z WARN  [BlockingBatchedESOutput] Error while waiting for healthy Elasticsearch cluster. Not flushing.
java.util.concurrent.TimeoutException: Elasticsearch cluster didn't get healthy within timeout
... ....
2015-11-23T09:24:45.833Z ERROR [Messages] Failed to index [1269] messages. Please check the index error log in your web interface for the reason. Error: failure in bulk execution:
[0]: index [graylog2_365], type [message], id [06d50732-91c4-11e5-be71-005056a37afa], message [ClusterBlockException[blocked by: [FORBIDDEN/8/index write (api)];]]
[1]: index [graylog2_365], type [message], id [06d52e40-91c4-11e5-be71-005056a37afa], message [ClusterBlockException[blocked by: [FORBIDDEN/8/index write (api)];]]
[2]: index [graylog2_365], type [message], id [06d57c61-91c4-11e5-be71-005056a37afa], message [ClusterBlockException[blocked by: [FORBIDDEN/8/index write (api)];]]
[3]: index [graylog2_365], type [message], id [06d57c62-91c4-11e5-be71-005056a37afa], message [ClusterBlockException[blocked by: [FORBIDDEN/8/index write (api)];]]
[4]: index [graylog2_365], type [message], id [06d5f191-91c4-11e5-be71-005056a37afa], message [ClusterBlockException[blocked by: [FORBIDDEN/8/index write (api)];]]

```

**Comment by razvanphp** [comment]
We had this today too, really strange, graylog (1.2.2) is mostly unresponsive, I cannot even rotate the index..

**Comment by simmcomm** [comment]
This issue seems to be similar as #1609 

We lately also lost all journal messages (running Graylog 1.3.3) on ES 1.7.5, in which incident Graylog could not write to ES (but assumed so) and flushed all journal messages to NIL. In this latest case we had an issue that the current write index was closed due to a bug in an external maintenance script. However we had cases before where our 2 ES-Cluster had to restart, having it's status shortly in RED and longer period in YELLOW (until all the re-sync took place) and during that time frame GL tried to flush the journal to ES, which ES didn't accept.

I can't say if this only happened during the RED period or if this issue also happens during ES YELLOW state.

In any event Graylog should only remove messages from the message journal if it got a write ACK from ES. Otherwise the journal doesn't make much sense if we can't trust that it's reliably flushed into ES.

At this moment the only work-around is to manually stop message processing as soon as an administrator detects problems with ES and manually resume it once ES turns back GREEN.

**Comment by bernd** [comment]
This will be fixed with the upcoming Graylog 2.3 where we switched to the HTTP transport where `ClusterBlockException` errors will trigger the retry logic instead of discarding the messages.

**Comment by jingene** [comment]
@bernd Hi, Is this fixed #3927 PR?
